### PR TITLE
ADD: single video page, watch later.

### PR DESCRIPTION
**Pull request by zoef-yt**
[PREVIEW LINK](https://deploy-preview-5--game-plex-zoef.netlify.app/)
[FILES TO CHECK](https://github.com/zoef-yt/game-plex-react/pull/5/files)
In this PR, I added

- Watch later functionality
- Data for videos
- single video page
- changed animation for the appearance of the card
- Added infinite scrolling on suggested videos

### CODE REVIEW

**Review by rutu14 — APPROVED** [review]
The code is good to be merged.

File: src/Context/WatchLaterContext/WatchLaterContext.jsx [hunk 1]
```diff
@@ -4,11 +4,48 @@ import { useAxios } from '../../CustomHooks/useAxios';
...
+			setWatchLaterList(watchLaterResponse.watchlater);
+		} else {
+			getWatchLater();
+		}
+	}, [watchLaterResponse]);
+
+	useEffect(() => {
+		console.log(watchLaterError?.response);
+	}, [watchLaterError]);
+
```
Comment (itsmebipul00) [inline src/Context/WatchLaterContext/WatchLaterContext.jsx:22]: You could remove this

File: src/Context/WatchLaterContext/WatchLaterContext.jsx [hunk 2]
```diff
@@ -4,11 +4,48 @@ import { useAxios } from '../../CustomHooks/useAxios';
...
+			headers: { authorization: token },
+		});
+	};
+
+	const removeFromWatchLater = (videoId) => {
+		watchLaterServerCall({
+			method: 'delete',
+			url: `/api/user/watchlater/${videoId}`,
+			headers: { authorization: token },
+		});
```
Comment (itsmebipul00) [inline src/Context/WatchLaterContext/WatchLaterContext.jsx:45]: Appreciation:  Switching to useAxios

File: src/Pages/VideoListingPage/components/VideoCard.jsx [hunk 3]
```diff
@@ -1,23 +1,51 @@
...
-		const baseDelay = 40,
-			delayMultiplier = 80;
-		return baseDelay + index * delayMultiplier;
+		const baseDelay = 20,
+			delayMultiplier = 35,
+			maxAnimationRange = 1500;
+		return baseDelay + index * delayMultiplier > maxAnimationRange
+			? index * delayMultiplier - delayMultiplier
+			: baseDelay + index * delayMultiplier;
 	};
```
Comment (itsmebipul00) [inline src/Pages/VideoListingPage/components/VideoCard.jsx:13]: Appreciation: nicely done

**Review by itsmebipul00 — APPROVED** [review]
Awesome animations, nice code, good to merge

**Review by zoyatanweer — APPROVED** [review]
- code is clean
- very nicely written
- good to merge

File: src/Context/WatchLaterContext/WatchLaterContext.jsx [hunk 4]
```diff
@@ -4,11 +4,48 @@ import { useAxios } from '../../CustomHooks/useAxios';
...
-		{ _id: 'O-w-NKMowIY', title: 'Valorant Skin Challenge | Zoef Games', creator: 'Zoef Games', categoryName: 'Challenges', id: '1' },
-		{ _id: 'SzuoB4_RpT4', title: "Marvel's Avengers Review(*NO SPOLIERS*)", creator: 'Zoef Games', categoryName: 'Podcast', id: '2' },
-	]);
-	return <WatchLaterContext.Provider value={{ watchLaterList }}>{children}</WatchLaterContext.Provider>;
+	const [watchLaterList, setWatchLaterList] = useState([]);
+	const { response: watchLaterResponse, operation: watchLaterServerCall, loading: loadingWatch, error: watchLaterError } = useAxios();
+	const token = localStorage.getItem('token');
+
+	useEffect(() => {
+		if (watchLaterResponse != null && watchLaterResponse.watchlater) {
```
Comment (muskanbagrecha) [inline src/Context/WatchLaterContext/WatchLaterContext.jsx:12]: You can use optional chaining here.

File: src/Pages/SingleVideoPage/SingleVideoPage.css [hunk 5]
```diff
@@ -0,0 +1,146 @@
+.video-page-body {
+	min-height: 100%;
+	max-width: 80%;
+	width: 100%;
+	display: grid;
+	grid-template-columns: 80% 20%;
+	grid-template-areas: 'video-section suggestion-section';
+	margin: 0 auto 0 auto;
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.css:8]: Could directly use 0 auto

File: src/Pages/SingleVideoPage/SingleVideoPage.css [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		grid-template-rows: 1fr 40rem;
+		grid-template-columns: 1fr;
+		grid-template-areas:
+			'video-section'
+			'suggestion-section';
+	}
+
+	.video-player {
+		overflow: hidden;
+		position: relative;
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.css:143]: Dont understand why this is positioned relatively.

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import { SuggestedVideos } from './SuggestedVideos';
+import { VideoPlayer } from './VideoPlayer';
+import { AddToPlaylistIcon, FilledWatchLaterIcon, LikeFilledIcon, LikeIcon, WatchLaterIcon } from '../../Assets/svg/AllSVG';
+const SingleVideoPage = () => {
+	const { id: videoURLId } = useParams();
+	const { videos } = useVideo();
+	const [currentVideoDetail, setCurrentVideoDetail] = useState('');
+	useEffect(() => {
+		const videoDetail = videos.find((video) => video._id === videoURLId);
+		if (videoDetail !== -1) {
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.jsx:14]: MDN for find: The find() method returns the first element in the provided array that satisfies the provided testing function. If no values satisfy the testing function, [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) is returned.

Comment (muskanbagrecha) [reply src/Pages/SingleVideoPage/SingleVideoPage.jsx:14]: So wanted to know why you're checking with -1.

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+			setCurrentVideoDetail(videoDetail);
+		}
+		window.scrollTo(0, 0);
+	}, [videoURLId, videos]);
+
+	const randomlySortedVideos = [...videos].sort(() => Math.random() - 0.5);
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.jsx:20]: Neat trick to sort videos randomly.

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	const { watchLaterList, addToWatchLater, removeFromWatchLater } = useWatchLater();
+	const { user } = useAuth();
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.jsx:22]: If you have an authcontext then why are you using the encoded token directly from local storage in your watch later context?

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+									<div className='cta-btn'>
+										{' '}
+										<AddToPlaylistIcon /> Playlist
+									</div>
+								</div>
+							</div>
+
+							<div className='meta-data text-grey'>
+								<span className='views'>{currentVideoDetail.views} views </span>·
+								<span className='date'> {currentVideoDetail.date}</span>
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SingleVideoPage.jsx:59]: You can destructure props of currentVideoDetail.

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { useEffect, useState, useRef } from 'react';
+import './SingleVideoPage.css';
+import { useVideo } from '../../Context';
+import { VideoCard } from '../VideoListingPage/components/VideoCard';
+const SuggestedVideos = ({ videoURLId, randomlySortedVideos }) => {
+	const { videos } = useVideo();
+	const scrollBar = useRef();
+	const [suggestedVideo, setSuggestedVideo] = useState([]);
+	const [scrollPos, setScrollPos] = useState({ positionRightNow: 0, heightRemaining: 600 });
+	const videosLimit = 5;
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SuggestedVideos.jsx:10]: Better naming convention would be VIDEOLIMIT.

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+	const scrollHandler = (e) => {
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SuggestedVideos.jsx:12]: Learnt something new here.

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+		const target = e.target;
+		setScrollPos({ positionRightNow: target.scrollTop, heightRemaining: target.scrollHeight - target.clientHeight });
+	};
+
+	useEffect(() => {
+		scrollBar.current.scrollTop = 0;
+		const newMustWatchVideos = videos.slice(0, 7);
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SuggestedVideos.jsx:19]: What is this 7 here?

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	}, [videoURLId, videos]);
+
+	useEffect(() => {
+		if (videos.length - 1 > suggestedVideo.length && scrollPos.positionRightNow > scrollPos.heightRemaining - 200) {
+			const newData = [...videos].slice(suggestedVideo.length + 1, suggestedVideo.length + 1 + videosLimit);
+			setSuggestedVideo((prev) => [...prev, ...newData]);
+		}
+	}, [scrollPos]);
+
+	return (
```
Comment (muskanbagrecha) [inline src/Pages/SingleVideoPage/SuggestedVideos.jsx:31]: Appreciation: It's great that you pulled off infinite scroll feature.

**Review by muskanbagrecha — APPROVED** [review]
Great work on infinite scrolling, learnt something new. Everything else looks good. Do look into the suggested changes.

File: src/Context/WatchLaterContext/WatchLaterContext.jsx (hunk 1 above)
Comment (zoef-yt) [reply src/Context/WatchLaterContext/WatchLaterContext.jsx:22]: done

File: src/Context/WatchLaterContext/WatchLaterContext.jsx (hunk 4 above)
Comment (zoef-yt) [reply src/Context/WatchLaterContext/WatchLaterContext.jsx:12]: done

File: src/Pages/SingleVideoPage/SingleVideoPage.css (hunk 5 above)
Comment (zoef-yt) [reply src/Pages/SingleVideoPage/SingleVideoPage.css:8]: done

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx (hunk 13 above)
Comment (zoef-yt) [reply src/Pages/SingleVideoPage/SuggestedVideos.jsx:19]: Forgot to replace it with the variable, Thanks.

File: src/Pages/SingleVideoPage/SuggestedVideos.jsx (hunk 11 above)
Comment (zoef-yt) [reply src/Pages/SingleVideoPage/SuggestedVideos.jsx:10]: Done.

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx (hunk 8 above)
Comment (zoef-yt) [reply src/Pages/SingleVideoPage/SingleVideoPage.jsx:20]: Thanks

File: src/Pages/SingleVideoPage/SingleVideoPage.jsx (hunk 7 above)
Comment (zoef-yt) [reply src/Pages/SingleVideoPage/SingleVideoPage.jsx:14]: Fixed it.

**Comment by zoef-yt** [comment]
@muskanbagrecha and @itsmebipul00 done with all the suggested changes in #8 .

**Pull request merged**